src/reducers/vis-state-merger.js

```diff
@@ -111,13 +110,24 @@ export function mergeFilters(state, filtersToMerge) {
   // flatten all filter dataIds
   const datasetsToFilter = uniq(flattenDeep(merged.map(f => f.dataId)));
 
-  const filtered = applyFiltersToDatasets(datasetsToFilter, updatedDatasets, updatedFilters);
+  const filtered = applyFiltersToDatasets(datasetsToFilter, updatedDatasets, updatedFilters, state.layers);
+
+  const features = updatedFilters.reduce((acc, f) => {
+    return f.type === FILTER_TYPES.polygon ? [...acc, f.value] : acc;
+  }, []);
 
   return {
     ...state,
     filters: updatedFilters,
     datasets: filtered,
-    filterToBeMerged: unmerged
+    filterToBeMerged: unmerged,
```

Comment:
now that we merge filter twice, `filterToBeMerged` should be `[...state.filterToBeMerged, ...unmerged]`

Reply:
this change is breaking a lot of tests. need to dig a little bit

Reply:
this may not work because if we consider the following scenario:
- a filter is added to filterToBeMerged list
- a new dataset is added that matches such filter, therefore, we try to merge filters again
- when we merge filter, ```state.filterToBeMerged``` has one element and unmerged has 0, therefore, the new state ```filterToBeMerged``` property will have again one element because of ```[...state.filterToBeMerged, ...unmerged]```

Reply:
@heshan0131 ^

Reply:
in cases where filterA is regular filter, filterB is layer based filter.  both are waiting for the dataset to be loaded. when we run mergeFilter once, we save filterA to unmerged. and when we run it again, we clear out unmerged, and save filterB to unmerged. It's wrong, because we are losing filterA

Reply:
got it. Need to review the logic and understand why a lot of tests are failing